src/sync/app.cpp

```diff
@@ -0,0 +1,188 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or utilied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "app.hpp"
+#include <sstream>
+#include "sync_manager.hpp"
+#include "generic_network_transport.hpp"
+#include "app_credentials.hpp"
+#include "../external/json/json.hpp"
+
+// wrap an optional json key into the Optional type
+#define WRAP_JSON_OPT(JSON, KEY, RET_TYPE) \
+JSON.find(KEY) != JSON.end() ? Optional<RET_TYPE>(JSON[KEY].get<RET_TYPE>()) : realm::util::none
+
+namespace realm {
+namespace app {
+
+std::shared_ptr<App> App::app(const std::string app_id,
```

Comment:
IMO this function just makes the code using it less understandable than calling `make_shared<App>()` directly would be (and there doesn't appear to be any reason the App needs to be wrapped in a shared_ptr at all in the test). We wrap the call to `make_shared()` in other places because of `enabled_shared_from_this()`, which makes it so that an instance of the type which isn't owned by a shared_ptr will crash at runtime, but that isn't used here.